src/redisearch_rs/slots_tracker/src/slot_set.rs

```diff
@@ -0,0 +1,781 @@
+            return CoverageRelation::Equals;
+        }
+
+        // Check if union covers all input slots
+        for &range in ranges {
+            if !union.range_is_covered(range.start, range.end) {
+                return CoverageRelation::NoMatch;
+            }
+        }
+
+        // Union covers input but has extra slots
+        CoverageRelation::Covers
+    }
+
+    // ========================================================================
+    // Private helper methods:
+    // ========================================================================
+
+    /// Normalizes the internal ranges: sorts and merges overlapping/adjacent ranges.
+    fn normalize(&mut self) {
+        if self.ranges.len() <= 1 {
+            return;
+        }
+
+        // Sort by start position
+        self.ranges.sort_unstable_by_key(|r| r.start);
+
+        // Merge overlapping and adjacent ranges in-place
+        let mut write_pos = 0;
+
+        for read_pos in 1..self.ranges.len() {
+            let current = self.ranges[read_pos];
+
+            if self.ranges[write_pos].end + 1 >= current.start {
+                // Overlapping or adjacent - merge into write_pos
+                self.ranges[write_pos].end = self.ranges[write_pos].end.max(current.end);
+            } else {
+                // Not adjacent - advance write position and copy
+                write_pos += 1;
+                self.ranges[write_pos] = current;
+            }
+        }
+
+        // Truncate to the merged length
+        self.ranges.truncate(write_pos + 1);
+    }
+
+    /// Checks if a single range is completely covered by this set.
+    fn range_is_covered(&self, start: u16, end: u16) -> bool {
+        let mut current = start;
+
+        for &range in &self.ranges {
+            // Range starts after current position - gap found
+            if range.start > current {
+                return false;
+            }
+            // Range covers current position
+            // No need to check `range.start <= current` due to previous condition
+            if range.end >= current {
+                current = range.end + 1;
+                if current > end {
+                    // Covered everything
+                    return true;
+                }
+            }
+        }
+
+        // Didn't cover everything
+        false
+    }
+}
+
+// Implement PartialEq with slices for convenient comparisons
+impl PartialEq<[SlotRange]> for SlotSet {
+    fn eq(&self, other: &[SlotRange]) -> bool {
+        self.ranges == other
+    }
+}
+
+impl PartialEq<&[SlotRange]> for SlotSet {
+    fn eq(&self, other: &&[SlotRange]) -> bool {
+        self.ranges == *other
+    }
+}
+
+impl PartialEq<SlotSet> for [SlotRange] {
+    fn eq(&self, other: &SlotSet) -> bool {
+        self == other.ranges
+    }
+}
+
+impl PartialEq<SlotSet> for &[SlotRange] {
+    fn eq(&self, other: &SlotSet) -> bool {
+        *self == other.ranges
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
is there a way to parametrize these tests? so that we can run the same test with different inputs (like contiguous ranges, previously normalized or not, etc ...?)